### ISSUE TITLE
Uniqueness validator succeeds on UPDATE with non-unique values if model has columnMap

### ISSUE THREAD

**Issue by JetA2** [phalcon/cphalcon#13398]
When a model uses a Uniqueness validator, the validation succeeds for non-unique values when an UPDATE is performed and the model uses a columnMap.

I believe this is due to the validator not using the mapped column name when creating the part of the query that excludes the current object from validation. See the line of code below:

https://github.com/phalcon/cphalcon/blob/65183349743e899977adc643f103b887612ffe98/phalcon/validation/validator/uniqueness.zep#L298

Sample code to reproduce:
```php
<?php

$di = new Phalcon\DI\FactoryDefault();

$di->setShared('db', function () {
    return new \Phalcon\Db\Adapter\PDO\Sqlite(array(
        'dbname'   => ':memory:'
    ));
});

class Test extends \Phalcon\Mvc\Model
{
    public $mappedPrimaryKey;
    public $mappedData;

    public function columnMap()
    {
        // Keys are the real names in the table and
        // the values their names in the application
        //
        return array(
            "primaryKey"    => "mappedPrimaryKey",  
            "data"          => "mappedData"
        );
    }

    public function validation()
    {
        $validator = new Phalcon\Validation();          

            $validator->add( "mappedData",
                new Phalcon\Validation\Validator\Uniqueness( array(
                    "message" => "Data must be unique"
         )));

        return $this->validate( $validator );
    }
}

use \Phalcon\Db\Column as Column;

$di['db']->createTable(
    'test',
    null,
    [
       'columns' => [
            new Column(
                'primaryKey',
                [
                    'type'          => Column::TYPE_INTEGER,
                    'notNull'       => true,
                    'autoIncrement' => true,
                    'primary'       => true,
                ]
            ),
            new Column(
                'data',
                [
                    'type'    => Column::TYPE_TEXT,
                    'notNull' => true,
                ]
            )
        ]
    ]
);

$testObject1 = new Test;
$testObject1->mappedData = "Object 1";

$testObject1->save();

$testObject2 = new Test;
$testObject2->mappedData = "Object 2";

$testObject2->save();

$allRecords = Test::find();
foreach ( $allRecords as $record )
{
    echo $record->mappedPrimaryKey . ": " . $record->mappedData . "\n";
}
echo "\n";

$result = Test::findFirst( "mappedPrimaryKey = 1" );
$result->mappedData = "Object 2";   // Not unique
$result->save();

$allRecords = Test::find();
foreach ( $allRecords as $record )
{
    echo $record->mappedPrimaryKey . ": " . $record->mappedData . "\n";
}

echo "\nPhalcon Version: ", Phalcon\Version::get(), "\n";
```

My output:
```
1: Object 1
2: Object 2

1: Object 2
2: Object 2

Phalcon Version: 3.4.0
```

**Comment by Jurigag** [comment]
The only way i see it it's somehow problem with sqlite adapter or something. What about mysql?

**Comment by JetA2** [comment]
From what I can see in the source code, there is an issue with the validator itself. The validator uses the readAttribute function of the Phalcon\Mvc\Model class to fetch the value of the primary key to bind in the query.

This is the readAttribute function:
```php
	/**
	 * Reads an attribute value by its name
	 *
	 * <code>
	 * echo $robot->readAttribute("name");
	 * </code>
	 */
	public function readAttribute(string! attribute)
	{
		if !isset this->{attribute} {
			return null;
		}

		return this->{attribute};
	}
```

If understand Zephir correctly, the readAttribute function is looking for a member variable with the name supplied to the function in the "attribute" parameter. On the model in the sample code, this member variable would be named "mappedPrimaryKey" since the model uses a column map.

However, the uniqueness validator supplies the readAttribute function with the variable "primaryField":
```php
let params["bind"][] = record->readAttribute(primaryField);
```

The variable "primaryField" is populated from the function "getPrimaryKeyAttributes" from the Phalcon\Mvc\Model\MetaData class. The function gets the primary key column name from the database schema. Thus the variable "primaryField" in the validator will contain the column name "primaryKey" for the sample code.

The "readAttribute" function will not find a member variable in the model named "primaryKey" since the member variable is named "mappedPrimaryKey" and will return null instead of the numeric primary key value. A null value will then be bound in the query, which can be seen in the query log.

If line 298 in the uniqueness validator is changed to the following I believe it should take care of the issue:
```php
let params["bind"][] = record->readAttribute( this->getColumnNameReal(record, primaryField) );
```

**Comment by Jurigag** [comment]
That's weird, in deed it returns null. Not sure why someone missed this.

**Comment by Jurigag** [comment]
Well i guess this almost never happened because you almost always have primary key like id in table and in model $id as well and that's why no one had this issue.

**Comment by sergeyklay** [comment]
Fixed in the `3.4.x` branch. Feel free to open a new issue if the problem appears again. Thank you for contributing.
